rand_core/src/lib.rs

```diff
@@ -212,6 +188,81 @@ pub trait RngCore {
 /// [`BlockRngCore`]: block::BlockRngCore
 pub trait CryptoRng: RngCore {}
 
+/// The potentially fallible core of a random number generator.
+pub trait TryRngCore {
+    /// The type returned in the event of a RNG error.
+    type Error: fmt::Debug + fmt::Display;
```

Comment:
Given an `rng: R` where `R: TryRngCore`, the only things we can do with the error are to format it to a message. Is this problematic for `no_std` where formatting machinery may not be available?

Also, this makes `TryRngCore` nearly useless for `dyn` usages (the error type would have to be tied to that of a single RNG), but this may not matter.

We could maybe mitigate both issues if we keep an `enum ErrorKind` which `Error` must convert into and add a converting wrapper, though I don't know if it's especially useful. It would be a breaking change to add later however.

Reply:
>Is this problematic for no_std where formatting machinery may not be available?

AFAIK formatting machinery is always available on `no_std`. Though it often makes sense to avoid it, since, unfortunately, it can add a significant amount of binary bloat.

>this makes TryRngCore nearly useless for dyn usages

Yes, this is why we could add the `DynTryRng` trait to amend this. But users could do define a similar trait themselves and we also could do it later in a backwards compatible manner.

Reply:
It may be my bias talking, but I think that most users of this trait would be fine with simply bubbling the generic error. Usually, meaningful error processing will be possible only when you know a concrete RNG type and it does not look like `ErrorKind` would be useful in practice.

Imagine a crypto library which allows users to select primary RNG (e.g. it may be `OsRng` or a certified HW-based RNG). Such library can not do any meaningful RNG error processing outside of bubbling the error or unwrapping it.

Reply:
Fair argument :+1: